nickel/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+                let mut ansi;
+                let mut no_color;
+                let writer: &mut dyn WriteColor = match color {
+                    Color::Ansi => {
+                        ansi = Ansi::new(write);
+                        &mut ansi
+                    }
+                    Color::Off => {
+                        no_color = NoColor::new(write);
+                        &mut no_color
+                    }
+                };
+                diagnostics.diagnostics.iter().try_for_each(|d| {
+                    Ok(codespan_reporting::term::emit(
+                        writer,
+                        &codespan_reporting::term::Config::default(),
+                        &self.files,
+                        d,
+                    )?)
+                })
+            }
+            ErrorFormat::Json => Ok(serde_json::to_writer(write, &diagnostics)?),
+            ErrorFormat::Yaml => Ok(serde_yaml::to_writer(write, &diagnostics)?),
+            ErrorFormat::Toml => Ok(write.write_all(toml::to_string(&diagnostics)?.as_bytes())?),
+        }
+    }
+}
+
+impl From<std::io::Error> for Error {
+    fn from(e: std::io::Error) -> Self {
+        Error {
+            error: Box::new(nickel_lang_core::error::Error::IOError(e.into())),
+            files: Files::new(),
+        }
+    }
+}
+
+impl From<nickel_lang_core::error::ExportError> for Error {
+    fn from(e: nickel_lang_core::error::ExportError) -> Self {
+        Error {
+            error: Box::new(e.into()),
+            files: Files::new(),
+        }
+    }
+}
+
+impl Context {
+    /// Set the interpreter's search path for imports.
+    ///
+    /// When importing a file, Nickel searches for it relative to the file doing the
+    /// import. If not found, it will search in the paths provided here.
+    ///
+    /// Note that unlike the Nickel command line program, the embedded Nickel
+    /// interpreter does not automatically honor the `NICKEL_IMPORT_PATH`
+    /// environment variable. If you want to support `NICKEL_IMPORT_PATH`, you
+    /// should read the environment variable yourself and pass the resulting
+    /// paths to this method.
+    pub fn with_import_paths(self, import_paths: Vec<OsString>) -> Self {
+        Context {
+            import_paths,
+            ..self
+        }
+    }
+
+    /// Provide a destination for the output of `std.trace`.
+    ///
+    /// If you don't provide a destination, `std.trace` will have
+    /// no effect.
+    pub fn with_trace(self, trace: Trace) -> Self {
+        Context { trace, ..self }
+    }
+
+    /// Provide a name for the main input program.
+    ///
+    /// This is used to format error messages. If you read the main input
+    /// program from a file, its path is a good choice.
+    pub fn with_source_name(self, name: String) -> Self {
+        Context {
+            name: Some(name),
+            ..self
+        }
+    }
+
+    /// Evaluate a Nickel program deeply, returning the resulting expression.
```

Comment:
As a public API, it might be helpful to detail just quickly what does "deep evaluation" mean. Stating that by default we don't recurse into containers (lazy evaluation), and give an example with a record and a failing contract that is not triggered by lazy evaluation, but is by deep evaluation, or something like that?

Reply:
Ok, I saw `eval_shallow` afterwards. So there are enough details there, but I still think we should define idiosyncratic jargon like "deep". If only that as opposed to `eval_shallow`, this recurse into records and arrays.